File: seminar-04/binary_search/python/binary_search_bounds.py
# lbound: первый индекс i, такой что a[i] >= target
# возвращает значение в диапазоне [0..n]
def lbound(target, elems):
    l = -1
    r = len(elems)
    while r - l > 1:
        m = (l + r) // 2
        if elems[m] >= target:
            r = m
        else:
            l = m
    return r

# rbound: последний индекс i, такой что a[i] <= target
# возвращает значение в диапазоне [-1..n-1]
# (эквивалент: (первый индекс > target) - 1)
def rbound(target, elems):
    l = -1
    r = len(elems)
    while r - l > 1:
        m = (l + r) // 2
        if elems[m] > target:
            r = m
        else:
            l = m
    return l

File: seminar-04/binary_search/cpp/binary_search_bounds.cpp
#include <vector>
using namespace std;

// lbound: первый индекс i, такой что a[i] >= target
// возвращает значение в диапазоне [0..n]
int lbound(const vector<int>& a, int target) {
    int l = -1;
    int r = (int)a.size();
    while (r - l > 1) {
        int m = l + (r - l) / 2;
        if (a[m] >= target) {
            r = m;
        } else {
            l = m;
        }
    }
    return r;
}

// rbound: последний индекс i, такой что a[i] <= target
// возвращает значение в диапазоне [-1..n-1]
int rbound(const vector<int>& a, int target) {
    int l = -1;
    int r = (int)a.size();
    while (r - l > 1) {
        int m = l + (r - l) / 2;
        if (a[m] > target) {
            r = m;
        } else {
            l = m;
        }
    }
    return l;
}
